### PR TITLE
AF-1251: Support to experimental features

### PR DESCRIPTION
**Pull request by pefernan**
AF-1538: Experimental Features Editor Screen
AF-1443: Global Experimental Features
AF-1442: Menus should support experimental features
AF-1436: API for Experimental LayoutDragComponent
AF-1449: Create a default view to show when displaying a disabled experimental feature
AF-1524: Add @ExperimentalFeature support for ResourceType
AF-1440: Add @ExperimentalFeature support for Editors.
AF-1438: Experimental Feature Annotation
AF-1439: Add @ExperimentalFeature support for Screens
AF-1439: Add @ExperimentalFeature support for Screens
AF-1441: Add @ExperimentalFeature support for Perspectives
AF-1445: Experimental Feature Core API

@ederign @tomasdavidorg could you please take a look?

This is a part of an ensemble, please merge with:
https://github.com/kiegroup/appformer/pull/475
https://github.com/kiegroup/kie-wb-common/pull/2138
https://github.com/kiegroup/kie-wb-distributions/pull/819
https://github.com/kiegroup/drools-wb/pull/940
https://github.com/kiegroup/optaplanner-wb/pull/303
https://github.com/kiegroup/kie-docs/pull/1131

### CODE REVIEW

**Comment by pefernan** [comment]
jenkins retest this

**Comment by pefernan** [comment]
jenkins retest this

File: uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/AbstractExperimentalFeaturesStorage.java [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+        if (!existsOnVFS || requiresVFSSync) {
+            storeFeatures(registeredFeatures);
+        }
+
+        return registeredFeatures;
+    }
+
+    private boolean syncLoadedFeatures(final List<ExperimentalFeatureImpl> registeredFeatures) {
+        final Collection<ExperimentalFeatureDefinition> expectedDefinitions = getSupportedDefinitions();
```
Comment (ederign) [inline uberfire-experimental/uberfire-experimental-backend/src/main/java/org/uberfire/experimental/service/storage/impl/AbstractExperimentalFeaturesStorage.java]: TODO

**Comment by pefernan** [comment]
@tomasdavidorg what do you think now?

**Comment by pefernan** [comment]
@tomasdavidorg again!

**Comment by pefernan** [comment]
jenkins retest this

File: uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/megamenu/WorkbenchMegaMenuProducerTest.java [hunk 2]
```diff
@@ -162,6 +170,24 @@ public void checkObservedEventsCallsPresenterStandaloneMode() {
...
+    public void testNotifyVisibilityChangeStandaloneMode() {
+        testNotifyVisibilityChange(true);
+    }
+
+    private void testNotifyVisibilityChange(boolean isStandalone) {
+        final WorkbenchMegaMenuPresenter presenter = getMegaMenuPresenter(isStandalone);
+
+        presenter.onPerspectiveVisibilityChange(new PerspectiveVisibiltiyChangeEvent("perspectiveId", false));
+
+        verify(presenter, times(1)).onPerspectiveVisibilityChange(any());
```
Comment (tomasdavidorg) [inline uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/megamenu/WorkbenchMegaMenuProducerTest.java]: One more `times(1)`.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/test/java/org/uberfire/ext/layout/editor/client/LayoutComponentPalettePresenterTest.java [hunk 3]
```diff
@@ -62,104 +116,123 @@ public void testInitialization() {
...
-        verify(view).addDraggableComponentGroup(any());
-        assertEquals(1, presenter.getLayoutDragComponentGroups().size());
-        assertNotNull(presenter.getLayoutDragComponentGroups().get(DRAGGABLE_GROUP_NAME));
+    public void testAddAllDraggableGroups() {
+
+        presenter.addDraggableGroups(Arrays.asList(new Group1LayoutComponentPaletteGroupProvider(true), new Group2LayoutComponentPaletteGroupProvider(), new Group3LayoutComponentPaletteGroupProvider()));
+
+        verify(dragComponentGroupPresenter, times(3)).init(any());
+        verify(dragComponentGroupPresenter, times(3)).getView();
+        verify(dragComponentGroupView, times(1)).setExpanded(true);
```
Comment (tomasdavidorg) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/test/java/org/uberfire/ext/layout/editor/client/LayoutComponentPalettePresenterTest.java]: One more `times(1)`.

File: uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/test/java/org/uberfire/ext/layout/editor/client/LayoutComponentPalettePresenterTest.java [hunk 4]
```diff
@@ -62,104 +116,123 @@ public void testInitialization() {
...
+        disabledExperimentalFeatures.add(Group1LayoutDragComponent3.class.getName());
+        disabledExperimentalFeatures.add(Group2LayoutComponentPaletteGroupProvider.class.getName());
+        disabledExperimentalFeatures.add(Group3LayoutDragComponent1.class.getName());
+
+        presenter.addDraggableGroups(Arrays.asList(new Group1LayoutComponentPaletteGroupProvider(true), new Group2LayoutComponentPaletteGroupProvider(), new Group3LayoutComponentPaletteGroupProvider()));
+
+        verify(dragComponentGroupPresenter, times(2)).init(any());
+        verify(dragComponentGroupPresenter, times(2)).getView();
+        verify(dragComponentGroupView, times(1)).setExpanded(true);
+        verify(dragComponentGroupView, times(1)).setExpanded(false);
```
Comment (tomasdavidorg) [inline uberfire-extensions/uberfire-layout-editor/uberfire-layout-editor-client/src/test/java/org/uberfire/ext/layout/editor/client/LayoutComponentPalettePresenterTest.java]: Two more `times(1)`.

File: uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/WorkbenchMenuBarPresenterTest.java [hunk 5]
```diff
@@ -361,6 +364,43 @@ public void testMenuInsertionOrder() {
...
+        final Menus menus = MenuFactory.newSimpleItem(NAME).withItems(items).endMenu().build();
+
+        when(experimentalActivitiesAuthorizationManager.authorizeActivityId(anyString()))
+                .thenAnswer((Answer<Boolean>) invocationOnMock -> !SECOND_PERSPECTIVE_ID.equals(invocationOnMock.getArguments()[0]));
+
+        presenter.addMenus(menus);
+
+        verify(experimentalActivitiesAuthorizationManager, times(3)).authorizeActivityId(anyString());
+        verify(view, times(2)).setMenuItemVisible(anyString(), eq(true));
+        verify(view, times(1)).setMenuItemVisible(anyString(), eq(false));
```
Comment (tomasdavidorg) [inline uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/WorkbenchMenuBarPresenterTest.java]: One more `times(1)`.

File: uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/WorkbenchMenuBarPresenterTest.java [hunk 6]
```diff
@@ -361,6 +364,43 @@ public void testMenuInsertionOrder() {
...
+    }
+
+    @Test
+    public void testShowHideExperimentalPerspectiveMenus() {
+        testAddMenusWithExperimentalPerspective();
+
+        presenter.onPerspectiveVisibilityChange(new PerspectiveVisibiltiyChangeEvent(PERSPECTIVE_ID, false));
+        verify(view, times(1)).setMenuItemVisible(PERSPECTIVE_ID, false);
```
Comment (tomasdavidorg) [inline uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/WorkbenchMenuBarPresenterTest.java]: Three more `times(1)`.

File: uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/megamenu/WorkbenchMegaMenuPresenterTest.java [hunk 7]
```diff
@@ -742,6 +755,45 @@ public void setupHomeLinkWithPermissionToAccessDefaultPerspective() {
...
+
+        when(experimentalActivitiesAuthorizationManager.authorizeActivityId(anyString()))
+                .thenAnswer((Answer<Boolean>) invocationOnMock -> !SECOND_PERSPECTIVE_ID.equals(invocationOnMock.getArguments()[0]));
+
+        presenter.addMenus(menus);
+
+        verify(experimentalActivitiesAuthorizationManager, times(3)).authorizeActivityId(anyString());
+        verify(presenter.canHideMenuItemByIdentifier, times(3)).put(any(), any());
+        verify(presenter.canHideMenuItemByIdentifier.get(PERSPECTIVE_ID), times(2)).show();
+        verify(presenter.canHideMenuItemByIdentifier.get(SECOND_PERSPECTIVE_ID), times(1)).hide();
```
Comment (tomasdavidorg) [inline uberfire-workbench/uberfire-workbench-client/src/test/java/org/uberfire/client/workbench/widgets/menu/megamenu/WorkbenchMegaMenuPresenterTest.java]: One more `times(1)`.

**Comment by pefernan** [comment]
jenkins retest this

**Comment by ederign** [comment]
@pefernan could you please fix the conflicts of: https://github.com/kiegroup/kie-wb-distributions/pull/819#pullrequestreview-162576337

**Comment by ederign** [comment]
@porcelli do you want to take a look on this?

**Comment by ederign** [comment]
jenkins retest this

**Comment by ederign** [comment]
please jenkins execute full downstream build

**Comment by ederign** [comment]
We have a green light as soon as https://github.com/kiegroup/kie-docs/pull/1131 is approved.

**Comment by pefernan** [comment]
please jenkins execute full downstream build

**Comment by pefernan** [comment]
@ederign @tomasdavidorg downstream build failed because of the unstable tests we do know, but wars were succesfully generated.

**Comment by pefernan** [comment]
please jenkins execute full downstream build

**Comment by pefernan** [comment]
please jenkins execute full downstream build

**Comment by pefernan** [comment]
please jenkins execute full downstream build

**Pull request merged**